Bugfix/product image loader

przemyslawspaczek opened the pull request:
### Short description and why it's useful
I extract to the new component images loading strategy implemented in Product Gallery Carousel and use it on Producta Gellery Carousel and Product Gallery Zoom Carousel components.

### Which environment this relates to
Check your case. In case of any doubts please read about [Release Cycle](https://docs.vuestorefront.io/guide/basics/release-cycle.html)

- [x] Test version (https://test.storefrontcloud.io) - this is a new feature or improvement for Vue Storefront. I've created branch from `develop` branch and want to merge it back to `develop`
- [ ] RC version (https://next.storefrontcloud.io) - this is a stabilisation fix for Release Candidate of Vue Storefront. I've created branch from `release` branch and want to merge it back to `release`
- [ ] Stable version (https://demo.storefrontcloud.io) - this is an important fix for current stable version. I've created branch from `hotfix` or `master` branch and want to merge it back to `hotfix`

### Upgrade Notes and Changelog

- [x] No upgrade steps required (100% backward compatibility and no breaking changes)
- [ ] I've updated the [Upgrade notes](https://github.com/DivanteLtd/vue-storefront/blob/develop/doc/Upgrade%20notes.md) and [Changelog](https://github.com/DivanteLtd/vue-storefront/blob/develop/CHANGELOG.md) on how to port existing VS sites with this new feature

**IMPORTANT NOTICE** - Remember to update `CHANGELOG.md` with description of your change

### Contribution and currently important rules acceptance
<!-- Please get familiar with following info -->

- [x] I read and followed [contribution rules](https://github.com/DivanteLtd/vue-storefront/blob/master/CONTRIBUTING.md)
- [x] I read the [TypeScript Action Plan](https://github.com/DivanteLtd/vue-storefront/blob/master/doc/TypeScript%20Action%20Plan.md) and adjusted my PR according to it
- [x] I read about [Vue Storefront Modules](https://github.com/DivanteLtd/vue-storefront/blob/master/doc/api-modules/about-modules.md) and I am aware that every new feature should be a module

pkarw commented on src/themes/default/components/core/ProductImage.vue:
```diff
@@ -0,0 +1,105 @@
+      return !this.showLowerQualityImage && !this.showHighQualityImage
+    },
+    showLowerQualityImage () {
+      return !!this.lowerQualityImageLoad && !this.highQualityImageLoad && !this.hidden
```
Isn’t the “!!” a typo?

przemyslawspaczek replied:
In this case "!!" is not necessary becouse I have everywhere have Boolen and no needed conver variable to Boolen. Remove it?

pkarw replied:
Yes Please. Please do also check the offline mode after this changes - should work without any issues as on demo

przemyslawspaczek replied:
Done. I find bug on initial valu but fixed it.

patzick commented on src/themes/default/components/core/ProductGalleryCarousel.vue:
```diff
@@ -16,41 +16,12 @@
-            ref="images"
-            @dblclick="openOverlay"
+            :images="images"
+            :hidden="hideImageAtIndex === index"
```
lets use v-show instead of hidden, because hiding these images is only for carousel purpose

patzick commented on src/themes/default/components/core/ProductGalleryCarousel.vue:
```diff
@@ -16,41 +16,12 @@
-            data-testid="productGalleryImage"
-            itemprop="image"
-          >
+            @dblclick.native="openOverlay"/>
```
please emit 'dblclick' event from child component, so we wouldn't have to use .native here

patzick commented on src/themes/default/components/core/ProductGalleryZoomCarousel.vue:
```diff
@@ -2,10 +2,13 @@
+          <product-image
+            :images="images"
+            :alt="productName | htmlDecode"
+            :hidden="hideImageAtIndex === index"
```
are you sure we want to hide images in this place?

patzick commented on src/themes/default/components/core/ProductGalleryZoomCarousel.vue:
```diff
@@ -2,10 +2,13 @@
+            :images="images"
+            :alt="productName | htmlDecode"
+            :hidden="hideImageAtIndex === index"
+            @click.native="navigate(key)"
```
also child component should emit `click` event so we don't have to use native here

patzick commented on src/themes/default/components/core/ProductGalleryZoomCarousel.vue:
```diff
@@ -27,15 +30,10 @@
-                data-testid="productGalleryImage"
-                itemprop="image"
-              >
+                :hidden="hideImageAtIndex === index" />
```
here also v-show instead of :hidden

patzick commented on src/themes/default/components/core/ProductImage.vue:
```diff
@@ -0,0 +1,105 @@
+      type: String,
+      required: true
+    },
+    hidden: {
```
we shouldn't have this prop here

patzick commented on src/themes/default/components/core/ProductImage.vue:
```diff
@@ -0,0 +1,105 @@
+    },
+    alt: {
+      type: String,
+      required: true
```
```suggestion
      default: ''
```
we shouldn't use required in props unless it's completely necessary, otherwise we should provide default values

patzick commented on src/themes/default/components/core/ProductImage.vue:
```diff
@@ -0,0 +1,105 @@
+import { onlineHelper } from '@vue-storefront/core/helpers'
+export default {
+  props: {
+    images: {
```
name suggests that it is an array, could you provide other name for that prop to avoid confusions?
Also make sure that prop has a default value (see comment below) and that if anything from product is not provided then it will at least display placeholder (for example there will be something wrong with this productImages object)

przemyslawspaczek commented:
@patzick 
- I removed hidden props from component and use v-show on parent component 
- Used $listeners to stop using '.native' event
- I set default for all props and removed required
- Cheged object name to 'image'

I used https://css-tricks.com/aspect-ratio-boxes/ to maintaining image size. I used  to this mathematic  formula {height% / ( width / 100 )}

patzick commented:
Merged and ready for tests

alinadivante commented:
@przspa You forgot about pictures in checkout :)
![image](https://user-images.githubusercontent.com/36628922/58699765-9cc8fd80-839e-11e9-8f22-206f8a13e871.png)

patzick commented:
Related to #2863 - @przspa please take a look :)

przspa commented:
> @przspa You forgot about pictures in checkout :)
> ![image](https://user-images.githubusercontent.com/36628922/58699765-9cc8fd80-839e-11e9-8f22-206f8a13e871.png) #3013 fixed my mistake
